### PR TITLE
 INTMDB-417-Network Container code change

### PR DESCRIPTION
**Pull request by ParthasarathyVaradhan**
## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s): 

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [✅] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ✅] This change requires a documentation update

## Required Checklist:

- [✅] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [✅] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [✅] I have run `make fmt` and formatted my code

## Further comments
Implemented Create, Read, Update, Delete, List methods for Network Container (initially only List and Delete were available).
Implemented Validator Code Changes.
Sample CFN template networkcontainer.sample-template.json
Input templates to create testing files for CFN invoke test

## Testing
Can be manually executed by using this json inputs and the command:

`cfn invoke resource {CREATE | READ | UPDATE | DELETE | LIST} file-name.json`

For CREATE AND LIST
```json
{
  "desiredResourceState" :  {
     "ProjectId": {GROUP-ID},
    "RegionName": {ATLAS-REGION-NAME},
    "AtlasCidrBlock": {ATLAS-CIDR-BLOCK} ,
    "ApiKeys" : {
      "PublicKey" : {PUBLIC-KEY},
      "PrivateKey" : {PRIVATE-KEY}
    }
     },
  "providerLogGroupName" : "",
  "previousResourceState": { }
}
```

For READ, UPDATE, DELETE
```json
{
  "desiredResourceState" :  {
    "Id":{CONTAINER-ID},
     "ProjectId": {GROUP-ID},
    "RegionName": {ATLAS-REGION-NAME},
     "AtlasCidrBlock": {ATLAS-CIDR-BLOCK} ,
    "ApiKeys" : {
      "PublicKey" : {PUBLIC-KEY},
      "PrivateKey" : {PRIVATE-KEY}
    }
  },
  "providerLogGroupName" : "",
  "previousResourceState": { }
}
```

### CODE REVIEW

**Review by govardhanpagidi — APPROVED** [review]
LGTM

File: cfn-resources/network-container/docs/README.md [hunk 1]
```diff
@@ -123,3 +87,11 @@ For more information about using the `Fn::GetAtt` intrinsic function, see [Fn::G
 
 Unique identifier of the Network Peering container.
 
+#### VpcId
+
+Unique identifier of the peer VPC.
```
Comment (Zuhairahmed) [inline cfn-resources/network-container/docs/README.md:92]:
```suggestion
Unique string that identifies the MongoDB Cloud VPC on AWS.
```

File: cfn-resources/network-container/docs/README.md [hunk 2]
```diff
@@ -123,3 +87,11 @@ For more information about using the `Fn::GetAtt` intrinsic function, see [Fn::G
...
+
+#### Provisioned
+
+Flag that indicates if the project has clusters deployed in the Network Peering container.
```
Comment (Zuhairahmed) [inline cfn-resources/network-container/docs/README.md:96]:
```suggestion
Boolean flag that indicates whether MongoDB Cloud clusters exist in the specified network peering container.
```

File: cfn-resources/network-container/docs/README.md [hunk 3]
```diff
@@ -59,37 +53,7 @@ _Type_: String
...
-_Required_: No
-
-_Type_: String
-
-_Update requires_: [No interruption](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-no-interrupt)
-
-#### AtlasCIDRBlock
+#### AtlasCidrBlock
 
 CIDR block that Atlas uses for your clusters.
```
Comment (Zuhairahmed) [inline cfn-resources/network-container/docs/README.md:58]:
```suggestion
IP addresses expressed in Classless Inter-Domain Routing (CIDR) notation that MongoDB Cloud uses for the network peering containers in your project. MongoDB Cloud assigns all of the project's clusters deployed to this cloud provider an IP address from this range. MongoDB Cloud locks this value if an M10 or greater cluster or a network peering connection exists in this project.

These CIDR blocks must fall within the ranges reserved per RFC 1918. AWS further limits the block to between the /24 and /21 ranges.

To modify the CIDR block, the target project cannot have:
- Any M10 or greater clusters
- Any other VPC peering connections

You can also create a new project and create a network peering connection to set the desired MongoDB Cloud network peering container CIDR block for that project. MongoDB Cloud limits the number of MongoDB nodes per network peering connection based on the CIDR block and the region selected for the project.

Example: A project in an Amazon Web Services (AWS) region supporting three availability zones and an MongoDB CIDR network peering container block of limit of /24 equals 27 three-node replica sets.
```

**Review by Zuhairahmed — COMMENTED** [review]
Going forward take as many definitions as you can can from Open API Spec: https://www.mongodb.com/docs/atlas/reference/api-resources-spec/#tag/Network-Peering-Containers/operation/createOneNewNetworkPeeringContainer

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 4]
```diff
@@ -3,46 +3,165 @@ package resource
...
+	if err != nil {
+		return handler.ProgressEvent{}, err
+	}
+
+	projectID := currentModel.ProjectId
+
+	containerRequest := &mongodbatlas.Container{}
+
+	if projectID == nil || *projectID == "" {
+		return handler.ProgressEvent{}, fmt.Errorf("error creating network container: `ProjectID` must be set")
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:45]:
```suggestion
		return handler.ProgressEvent{}, fmt.Errorf("error creating network container: `%s` must be set", constants.ProjectID )
```

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 5]
```diff
@@ -3,46 +3,165 @@ package resource
...
+	}
+
+	regionName := currentModel.RegionName
+	if regionName == nil || *regionName == "" {
+		return handler.ProgressEvent{}, fmt.Errorf("`error creating network container: RegionName` must be set")
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:50]:
```suggestion
		return handler.ProgressEvent{}, fmt.Errorf("`error creating network container: %s must be set", constants.RegionName)
```

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 6]
```diff
@@ -3,46 +3,165 @@ package resource
...
+	}
+	containerRequest.RegionName = *regionName
+	containerRequest.ProviderName = constants.AWS
+	CIDR := currentModel.AtlasCidrBlock
+	if CIDR == nil || *CIDR == "" {
+		return handler.ProgressEvent{}, fmt.Errorf("error creating network container: `AtlasCidrBlock` must be set")
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:56]: same

File: cfn-resources/network-container/cmd/resource/resource.go [hunk 7]
```diff
@@ -3,46 +3,165 @@ package resource
...
+	if err != nil {
+		if res.StatusCode == http.StatusConflict {
+			_, _ = logger.Debugf("Container already exists for this group. Try return existing container. err: %v", err)
+			containers, _, err2 := client.Containers.ListAll(context.Background(), *projectID, nil)
+			if err2 != nil {
+				_, _ = logger.Debugf("Error Containers.ListAll err:%v", err)
+				return handler.ProgressEvent{}, fmt.Errorf("error Containers.ListAll err:%v", err)
+			}
+			_, _ = logger.Debugf("containers:%v", containers)
+			first := containers[0]
```
Comment (andreaangiolillo) [inline cfn-resources/network-container/cmd/resource/resource.go:69]: [q] what if container is `nil`? I think we may want to check that container has one element

**Comment by govardhanpagidi** [comment]
@Zuhairahmed and @andreaangiolillo I have addressed the ReviewPoints raised in the PR, plz check

**Review by Zuhairahmed — APPROVED** [review]
LGTM

**Review by andreaangiolillo — APPROVED** [review]

**Pull request merged**